rust/src/asn1/mod.rs

```diff
@@ -0,0 +1,447 @@
+                    && data[0] & 0xC0 == 0
+                    && (obj.header.len > 256 || data.len() > 256)
+                {
+                    return Some(Asn1Check::DoubleOverflow);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn from_slice(input: &'a [u8], ad: &DetectAsn1Data) -> Result<Asn1<'a>, Asn1DecodeError> {
+        let mut results = Vec::new();
+        let mut rest = input;
+
+        // while there's data to process
+        while !rest.is_empty() {
+            let max_depth = ad.max_frames as usize;
+
+            if results.len() >= max_depth {
+                return Err(Asn1DecodeError::MaxFrames);
+            }
+
+            let res = parse_ber_recursive(rest, max_depth);
+
+            match res {
+                Ok((new_rest, obj)) => {
+                    results.push(obj);
+
+                    rest = new_rest;
+                }
+                // If there's an error, bail
+                Err(_) => {
+                    break;
+                }
+            }
+        }
+
+        Ok(Asn1(results))
+    }
+}
+
+/// Decodes Asn1 objects from an input + length while applying the offset
+/// defined in the asn1 keyword options
+fn asn1_decode<'a>(
+    buffer: &'a [u8],
+    buffer_offset: u32,
+    ad: &DetectAsn1Data,
+) -> Result<Asn1<'a>, Asn1DecodeError> {
+    // Get offset
+    let offset = if let Some(absolute_offset) = ad.absolute_offset {
+        absolute_offset
+    } else if let Some(relative_offset) = ad.relative_offset {
+        // relative offset in regards to the last content match
+
+        // buffer_offset (u32) + relative_offset (i32) => offset (u16)
+        u16::try_from({
+            if relative_offset > 0 {
+                buffer_offset
+                    .checked_add(u32::try_from(relative_offset)?)
+                    .ok_or(Asn1DecodeError::InvalidKeywordParameter)?
+            } else {
+                buffer_offset
+                    .checked_sub(u32::try_from(-relative_offset)?)
+                    .ok_or(Asn1DecodeError::InvalidKeywordParameter)?
+            }
+        })
+        .or(Err(Asn1DecodeError::InvalidKeywordParameter))?
+    } else {
+        0
+    };
+
+    // Make sure we won't read past the end or front of the buffer
+    if offset as usize >= buffer.len() {
+        return Err(Asn1DecodeError::InvalidKeywordParameter);
+    }
+
+    // Get slice from buffer at offset
+    let slice = &buffer[offset as usize..];
+
+    Asn1::from_slice(slice, ad)
+}
+
+/// Attempt to parse a Asn1 object from input, and return a pointer
+/// to the parsed object if successful, null on failure
+///
+/// # Safety
+///
+/// input must be a valid buffer of at least input_len bytes
+/// pointer must be freed using `rs_asn1_free`
+#[no_mangle]
+pub extern "C" fn rs_asn1_decode(
+    input: *const u8,
+    input_len: u16,
+    buffer_offset: u32,
+    ad_ptr: *const DetectAsn1Data,
+) -> *mut Asn1<'static> {
```

Comment:
I'm a bit surprised to see the lifetime here. Isn't `*mut Asn1` enough ?

Reply:
```rust
error[E0106]: missing lifetime specifier
   --> src/asn1/mod.rs:214:11
    |
214 | ) -> *mut Asn1 {
    |           ^^^^ help: consider giving it an explicit bounded or 'static lifetime: `Asn1 + 'static`
    |
    = help: this function's return type contains a borrowed value with an elided lifetime, but the lifetime cannot be derived from the arguments
```